File: LeetCode_每日一题/2023-05/01_1376. 通知所有员工所需的时间.go
/*
 * @Author: hakusai
 * @Date: 2023-05-21 19:49:46
 * @LastEditTime: 2023-05-21 19:57:18
 * @Description:
 */

package main

func numOfMinutes(n int, headID int, manager []int, informTime []int) int {
	g := make([][]int, n)
	for i, x := range manager {
		if x != -1 {
			g[x] = append(g[x], i)
		}
	}
	var dfs func(int) int
	dfs = func(i int) (ans int) {
		for _, j := range g[i] {
			ans = max(ans, dfs(j)+informTime[i])
		}
		return
	}
	return dfs(headID)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: LeetCode_每日一题/2023-05/01_1376. 通知所有员工所需的时间.java

/*
 * @Author: hakusai
 * @Date: 2023-05-21 19:49:53
 * @LastEditTime: 2023-05-21 19:51:15
 * @Description: 
 */
import java.util.*;
import java.io.*;
import java.math.*;

class Solution {
    private List<Integer>[] g;
    private int[] informTime;

    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        g = new List[n];
        Arrays.setAll(g, k -> new ArrayList<>());
        this.informTime = informTime;
        for (int i = 0; i < n; ++i) {
            if (manager[i] >= 0) {
                g[manager[i]].add(i);
            }
        }
        return dfs(headID);
    }

    private int dfs(int i) {
        int ans = 0;
        for (int j : g[i]) {
            ans = Math.max(ans, dfs(j) + informTime[i]);
        }
        return ans;
    }
}
